### ISSUE TITLE
Parsing header in MultipartReader does not support folding headers

### ISSUE THREAD

**Issue by Debbus72** [dotnet/aspnetcore#21282]
### Describe the bug
When parsing headers in the `Microsoft.AspNetCore.WebUtilities.MultipartReader` (src/Http/WebUtilities/src/MultipartReader.cs) the reader will incorrectly throw an InvalidDataException when the header is folded.

### To Reproduce
We have the following message-body that we want to read:
```

------=_Part_0_241408798.1588060137271
Content-Type: application/vnd.openxmlformats-officedocument.spreadsheetml.sheet; 
	name="1 of 2 test omgevingen.xlsx"
Content-Transfer-Encoding: binary
Content-Disposition: form-data; name="1 of 2 test omgevingen.xlsx"; filename="1 of 2 test omgevingen.xlsx"

...data...
------=_Part_0_241408798.1588060137271--
```
Notice that the Content-Type has a folding header (which is allowed in HTTP/1.1 for headers with media-types). However reading this will throw the following exception: 
`Invalid header line: 	name="1 of 2 test omgevingen.xlsx"`
```
System.IO.InvalidDataException:
   at Microsoft.AspNetCore.WebUtilities.MultipartReader+<ReadHeadersAsync>d__21.MoveNext (Microsoft.AspNetCore.WebUtilities, Version=3.1.3.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.WebUtilities.MultipartReader+<ReadNextSectionAsync>d__20.MoveNext (Microsoft.AspNetCore.WebUtilities, Version=3.1.3.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
```

### Further technical details
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.201\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

**Comment by davidfowl** [comment]
We don't support folding headers anywhere right now.

**Comment by samsosa** [comment]
> Historically, HTTP header field values could be extended over multiple lines by preceding each extra line with at least one space or horizontal tab (obs-fold).  This specification deprecates such line folding **except within the message/http media type**

See: [https://tools.ietf.org/html/rfc7230#section-3.2.4](https://tools.ietf.org/html/rfc7230#section-3.2.4)

... what a mess.

**Comment by Tratcher** [comment]
Multipart headers are a different beast from HTTP headers. They may look similar but they are governed by different specs and conventions. Multipart was developed for e-mails, and the folding was there because SMTP has line length limits. That's not relevant in HTTP.

What client is generating this format?

**Comment by Debbus72** [comment]
[SmartBear ReadyAPI](https://smartbear.com/product/ready-api/overview/).
It is the tool that our testers use for automated testing of our Apis. We just started with testing the POST of attachments and this surfaced when uploading Excel-files with larger content-types and larger attachment-names.

**Comment by Tratcher** [comment]
Honestly I'd file a bug with SmartBear. That format is rarely used in HTTP requests and we're unlikely to prioritize support for it

**Comment by Debbus72** [comment]
Yes, we already did. The reply we got was that they just follow the RFC... :-(

**Comment by Tratcher** [comment]
...in the wrong context. As a test client they should also compare their implementation to the other client's they're emulating like browsers which don't do this.

**Comment by shu-schleupen** [comment]
We are affected by exactly the same issue in our production environment. Our system is internet-facing and, thus, we cannot prevent the clients to send HTTP bodies without folded MIME headers. So, we would really appreciate a bugfix as soon as possible.

Although it is less common practice to use folding for MIME headers in an HTTP request, the InvalidDataException is still a bug in our understanding because the request body conforms to the MIME spec but is rejected by the MultipartReader. As was mentioned previously, header folding is deprecated for HTTP but for MIME it is not (at least we found that it is still allowed in the most recent MIME RFCs). The header which triggers the bug is located in the HTTP body and is therefore not governed by the HTTP spec but the MIME spec; for HTTP it is just text within the body.

**Comment by sja-schleupen** [comment]
"We don't support folding headers anywhere right now."
@davidfowl will MultipartReader conform to MIME-Spec and support folding (_not_ in HTTP-Headers) ?

**Comment by halter73** [comment]
I'm closing this because it's clearly unusual for clients sending "multipart/form-data" to rely on line folding, and being too lax parsing multipart data can lead to validation bypass attacks as described in https://blog.sicuranext.com/breaking-down-multipart-parsers-validation-bypass/.

While that doesn't discuss line folding specifically, https://www.akamai.com/blog/security/cve-2025-32094-http-request-smuggling shows how not being strict in disallowing line folding in headers can lead to real world issues. And while the Akamai issue was for HTTP/1.1 headers rather than multipart headers, it's the same syntax and therefore carries similar risks. Line folding inside the request body might not allow full blown request smuggling, but it definitely does seem like it could help enable validation bypass attacks if another parser doing validation interprets the folded header differently.

Also, it's a lot easier to work around MultipartReader not allowing line folding than it is to work around Kestrel not accepting line folding in the request headers because you always have the option to read the request body as a raw Stream. You could even create your own copy of MultipartRader that does allow line folding and use that if the scenario is critical for your application and you know the risks.

> > This specification deprecates such line folding **except within the message/http media type**

To be clear, this is referring to [message/http](https://datatracker.ietf.org/doc/html/rfc7230#section-8.3.1) Content-Type, not multipart/form-data.
